server/balancer.go

```diff
@@ -0,0 +1,316 @@
+		return true
+	}
+
+	return false
+}
+
+// Do implements Operator.Doop interface.
+func (lto *TransferLeaderOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	if lto.Check(req) {
+		return true, nil
+	}
+
+	// TODO: call admin command, maybe we can send it to a channel.
+	/*
+		res := &raft_cmdpb.AdminRequest{
+			CmdType: raft_cmdpb.AdminCmdType_TransferLeader.Enum(),
+			TransferLeader: &raft_cmdpb.TransferLeaderRequest{
+				Peer: lto.newLeader,
+			},
+		}
+	*/
+
+	return false, nil
+}
+
+// Balancer is an interface to select store regions for auto-balance.
+type Balancer interface {
+	Balance(cluster *ClusterInfo, c *raftCluster) (*BalanceOperator, error)
+}
+
+const (
+	// If the used ratio of one storage is greater than this value,
+	// it will never be used as a selected target and it should be rebalanced.
+	maxCapacityUsedRatio = 0.4
+)
+
+var (
+	_ Balancer = &capacityBalancer{}
+)
+
+type capacityBalancer struct {
+}
+
+func (cb *capacityBalancer) SelectFromStore(stores []*StoreInfo) *StoreInfo {
+	var resultStore *StoreInfo
+	for _, store := range stores {
+		if store == nil {
+			continue
+		}
+
+		if store.usedRatio() <= maxCapacityUsedRatio {
+			continue
+		}
+
+		if resultStore == nil {
+			resultStore = store
+			continue
+		}
+
+		if store.usedRatio() > resultStore.usedRatio() {
+			resultStore = store
+		}
+	}
+
+	return resultStore
+}
+
+func (cb *capacityBalancer) SelectToStore(stores []*StoreInfo) *StoreInfo {
+	var resultStore *StoreInfo
+	for _, store := range stores {
+		if store == nil {
+			continue
+		}
+
+		if resultStore == nil {
+			resultStore = store
+			continue
+		}
+
+		if store.usedRatio() < resultStore.usedRatio() {
+			resultStore = store
+		}
+	}
+
+	return resultStore
+}
+
+func (cb *capacityBalancer) SelectBalanceRegion(stores []*StoreInfo, regions *RegionsInfo) (*metapb.Region, *metapb.Peer) {
+	store := cb.SelectFromStore(stores)
+	if store == nil {
+		return nil, nil
+	}
+
+	// Random select one leader region from store.
+	storeID := store.store.GetId()
+	region := regions.randRegion(storeID)
+	leaderPeer := leaderPeer(region, storeID)
+	return region, leaderPeer
+}
+
+func (cb *capacityBalancer) SelectNewLeaderPeer(cluster *ClusterInfo, peers map[uint64]*metapb.Peer) *metapb.Peer {
+	stores := make([]*StoreInfo, 0, len(peers))
+	for storeID := range peers {
+		stores = append(stores, cluster.getStore(storeID))
+	}
+
+	store := cb.SelectToStore(stores)
+	if store == nil {
+		return nil
+	}
+
+	storeID := store.store.GetId()
+	return peers[storeID]
+}
+
+func (cb *capacityBalancer) Balance(cluster *ClusterInfo, c *raftCluster) (*BalanceOperator, error) {
+	// Firstly, select one balance region from cluster info.
+	stores := cluster.getStores()
+	region, oldLeader := cb.SelectBalanceRegion(stores, cluster.regions)
+	if region == nil || oldLeader == nil {
+		return nil, errors.New("Region cannot be found to do balance")
```

Comment:
An error returned or a nil value seems not different, you can ignore it outside.
You prefer to return nil value here?